spec/eof.md

```diff
@@ -161,10 +124,25 @@ Code executing within an EOF environment will behave differently than legacy cod
 - The instruction `JUMPDEST` is renamed to `NOP` and remains charging 1 gas without any effect.
     - Note: jumpdest-analysis is not performed anymore.
 - EOF contract may not deploy legacy code (it is naturally rejected on the code validation stage)
-- Legacy creation transactions (any tranactions with empty `to`) are invalid in case `data` contains EOF code (starts with `EF00` magic)
-- If instructions `CREATE` and `CREATE2` have EOF code as initcode (starting with `EF00` magic)
-    - deployment fails (returns 0 on the stack)
-    - caller's nonce is not updated and gas for initcode execution is not consumed
+
+#### Creation transactions
+
+Creation transactions (tranactions with empty `to`), with `data` containing EOF code (starting with `EF00` magic) are interpreted as having an EOF `initcontainer` in the `data` and:
+
+- intrinsic gas cost rules and limits defined in EIP-3860 for legacy creation transaction apply. The entire `initcontainer` is treated as initcode
+- **the initcontainer and all its subcontainers are validated recursively**
+- it is checked if the initcode container has its `len(data_section)` equal to `data_size`, i.e. data section content is exactly as the size declared in the header (see [Data section lifecycle](#data-section-lifecycle))
```

Comment:
To acomodate tooling I think we should consider data length re-writing in the same way that aux data is appended.

* if the container is the exact size of the data section no changes will be made
* If the container has a supplied data section that is less than specified in the header the data will be zero padded
* If there is more data after the end of the container, all if it will be added to the data section and the size of the data section will be updated.  If the container cannot be re-written (such as more than 64KiB of data) then the transaction will fail as though the container were invalid

Reply:
Is this comment related to the "no-TXCREATE" proposition in this PR or general? The rewriting of the data section and size during creation has not changed here, and is currently specced out slightly differently from what you've listed here. Refer to the (current) TXCREATE spec in the MegaEOF and the Data section lifecycle section (can also be EIP-7620). To address the specific scenarios from your bullet points:

>     if the container is the exact size of the input no changes will be made

This means that there is no static aux data, but there still might be dynamic aux data given in RETURNCONTRACT, and the data section will be expanded and its size increased

>     If the container is otherwise valid and has data section that is less than specified in the header the data will be zero padded. (or abort?)

In this case the data section must be expanded, aux data (static aux data) MUST be supplied to RETURNCONTRACT.

>     If there is more data after the end of the container all of it will be added to the data section and the size of the data section will be updated. If the container cannot be re-written (such as more than 64KiB of data) then the transaction will fail as though the container were invalid

This is never the case. If after parsing all the sections according to their declared sizes, there's still stray data in the container, this should not pass EOF validation.

Reply:
Can you clarify the distinction between the initcontainer (calling RETURNCONTRACT) and the deploy container (being referenced by RETURNCONTRACT), otherwise I'm not following the idea.

I would assume (and have put a mention of this in the new PR), that we can just interpret the bytes beyond the initcontainer (precisely, beyond the declared data_section of the initcontainer) as initcontainer's execution frame calldata, this seems like the most straightforward way, and the data_section rewriting only concerns the deploy contrainer, the initcontainer is left intact.